### ISSUE TITLE
nn.NLLLoss ignores class weights

### ISSUE THREAD

**Issue by idc9** [pytorch/pytorch#14289]
## 🐛 Bug

`nn.NLLLoss` and `nn.CrossEntropyLoss` ignore user provided class weights.

## To Reproduce

First run without class weights (gives correct answer)
``` python
import torch
import torch.nn as nn

log_prob = torch.tensor([[-0.0141, -4.2669]])
target = torch.tensor([0])
criterion = nn.NLLLoss()

criterion(log_prob, target)
```

```
tensor(0.0141)
```

Next run with class weights (gives same as above i.e. incorrect answer)
``` python
log_prob = torch.tensor([[-0.0141, -4.2669]])
target = torch.tensor([0])
weight = torch.tensor([2.0, 3.0])
criterion = nn.NLLLoss(weight=weight)

criterion(log_prob, target)
```

```
tensor(0.0141)
```
(should be 2 * 0.0141 = 0.0282)

## Expected behavior

The loss should be multiplied by the provided weight.

## Environment
PyTorch version: 0.4.1
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.13.6
GCC version: Could not collect
CMake version: version 3.12.0

Python version: 3.6
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip] Could not collect
[conda] pytorch                   0.2.0                py36_4cu75    soumith
[conda] torch                     0.4.1                     <pip>
[conda] torchvision               0.1.9                    py36_1    soumith
[conda] torchvision               0.2.1                     <pip>

**Comment by idc9** [comment]
Ok this is basically issue #14184 i.e.  `nn.NLLLoss(weight=weight, reduction='elementwise_mean')` ignores weights but  `nn.NLLLoss(weight=weight, reduction='sum')` does not ignore weights.  I'm not sure if this is a bug or a documentation issue, but I would expect (and want!) "elementwise_mean" to do a weighted mean.
